Add wrapper for MultiObjectiveOptimizer class and Pareto analysis

ml-evs opened the pull request:
See #3 for previous discussion.

ml-evs committed:
Added link to ACADO in README

ml-evs committed:
Wrapped some more DifferentialEquation constructors

ml-evs committed:
Added first pass at wrapping MultiObjectiveOptimizer class

ml-evs committed:
Added simple multiobjective example car_ws

ml-evs committed:
Added test for multi-objective OCP initialisation

dpinte committed:
FEAT: implement OCP::minimizeMayerTerm(idx, expression)

dpinte committed:
FEAT: support two PARETO_FRONT options

ml-evs commented:
Outstanding issues:

- car_ws example fails to converge, spitting out lots of ACADO warnings and errors.
- removing the call to solve in car_ws causes a segfault.

See https://github.com/force-h2020/acadopy/pull/3#issuecomment-491605170 for more detail.

dpinte commented on acadopy/bindings.pyx:
```diff
@@ -598,6 +627,7 @@ cdef class OptimizationAlgorithm:
 
     def __cinit__(self, ocp=None):
         cdef OCP _ocp
+        print('calling oa init')
```
I suggest we remove the print statements

dpinte commented on acadopy/bindings.pyx:
```diff
@@ -556,9 +573,18 @@ cdef class OCP:
+            expression = args[1]
+            self._thisptr.minimizeMayerTerm(multi_objective_index, deref(expression._thisptr))
+        elif len(args) == 1:
+            expression = args[1]
```
maybe add a comment here or a `log.debug` call to make sure the user is aware that this can cause segfaults.

dpinte commented on acadopy/bindings.pyx:
```diff
@@ -660,3 +697,36 @@ cdef class OptimizationAlgorithm:
+        elif _return_value != SUCCESSFUL_RETURN:
+            raise RuntimeError('ACADO optimizer failed.')
+
+    # def get_all_differential_states(self, filename):
```
remove commented code or uncomment it

dpinte commented on examples/car_ws.py:
```diff
@@ -0,0 +1,53 @@
+# (C) Copyright 2019 Enthought, Inc., Austin, TX
+# All rights reserved.
```
I would suggest adding a comment pointing to the original C++ example

dpinte reviewed (changes requested):
@ml-evs some minor comments

ml-evs committed:
FIX: Added guard to prevent double free when deallocating C++ objects.

- Whilst the `MultiObjectiveAlgorithm` class doesn't need to implement
the same destructor, it may be illustrative to future developers that
this is required when dellocating inherited objects.

dpinte commented on acadopy/bindings.pyx:
```diff
@@ -717,7 +718,9 @@ cdef class MultiObjectiveAlgorithm(OptimizationAlgorithm):
 
     def __dealloc__(self):
-        del self._thisptr
+        if self._owner:
```
I would suggest testing if the pointer is null in place of the ownership. Otherwise, the logic gets broken.

ml-evs replied:
Good idea, I'll do the same in the __dealloc__ for every class.

ml-evs committed:
Manually set pointers to null in every dealloc call

ml-evs committed:
Added catalyst mixing example

ml-evs committed:
Removed debug print statements

ml-evs committed:
Removed spurious(?) variable name from prototype function signature

ml-evs committed:
Partially fixed and added warning to minimizeMayerTerm function call

ml-evs committed:
TST: push more of car_ws example into tests

ml-evs committed:
Updated examples to more closely match ACADO versions

ml-evs committed:
TST: added wrapped examples as tests, excluding currently broken solve calls

ml-evs committed:
FIX: fixed typo in option settings that was breaking PARETO_FRONT_GENERATION option

ml-evs committed:
MAINT: tidied up global definition of options

ml-evs committed:
FIX: decoupled Python wrappers to OptimizationAlgorithm and MultiObjectiveAlgorithm

ml-evs commented:
Current status, and probably my final contribution:

1. Added the `multi_objective/catalyst_mixing_nbi` example to provide more debugging info for `car_ws` example.
2. The `car_ws` example now finishes a bit more cleanly, but still reports issues https://github.com/force-h2020/acadopy/pull/3#issuecomment-491605170 about invalid arguments. Looking at the actual values ACADO manages to spit out, they seem numerically very unstable (e.g. merit val in first step is 2.5e11 in Python, vs 1.8e4 in pure C++).
3. I've removed all inheritance between `OptimizationAlgorithm` and `MultiObjectiveAlgorithm` as this was causing issues when trying to call a method only implemented by the latter (`getAllDifferentialStates`). This means that there are identical implementations for `set()` and `solve()` in the bindings, but I don't see a way around it (with inheritance, the `self._thisptr` object has to be of type `OptimizationAlgorithm*`, which means the extra child method will fail to compile).
4. I think there might be something strange going on with the setting of options that is causing this to fail; unfortunately its quite hard to debug as there's no simple way to print out all the options used by ACADO (that I can find).

dpinte committed:
FEAT: restore inheritance between OptimizationAlgo and MultiObjectiveAlgo

dpinte committed:
TST: use a * import for now

dpinte commented:
@ml-evs I've restored the inheritance between OptimizationAlgorithm and MultiObjectiveAlgorithm (you can cast the pointer to the appropriate type to solve the problem you described above)

dpinte committed:
FEAT: clenaup inheritance patterns

dpinte committed:
TST: remove failing test that does not do what is expected

dpinte committed:
DEBUG: debugging the car ws example

dpinte committed:
FEAT: test the return value of calling subjectTo --> highlights a bug

dpinte committed:
FIX: fix constraint value in rocket flight control example

dpinte committed:
FIX: implement missing subjectTo support for ConstraintComponent with one argument

dpinte committed:
REF: cleanup class hierarchy

dpinte committed:
TST: fix rocket flight test and add constraint component test

dpinte commented:
@ml-evs We have stable results for the `car_ws` example. I found a number of small issues: 
- `OCP.subjectTo` was silently failing when adding a ConstraintComponent in place of DifferentialEquation.
- return values should be tested for all the calls to detect issues early
- the rocket flight example had a misconfigured constraint

dpinte committed:
DOC: add context and description of the car_ws example

dpinte committed:
FIX: misisng _owner attribute on the VariableGrid

dpinte committed:
TST: add a failing test for the Expression unary neg operator

dpinte committed:
FEAT: add solveSingleObject to MCO algorithm

dpinte committed:
TST: running example for the catalyst MCO

dpinte committed:
FEAT: add getParetoFront on the MCO

dpinte committed:
Add pareto front call to the catalyst_mixin_nbi example
